Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+
+At first glance, such a decision seems to violate the fundamental principles of .NET
+code running on any platform.  However, the intent is not for the vast majority of
+apps to use such optimizations.  The intended usage model is to allow library
+developers access to low level functions which enable optimization of key
+functions.  As such the use is expected to be limited, but performance critical.
+
+#### Intrinsic `IsSupported` granularity
+
+The existing implementation is using `IsSupported` to determine whether an existing
+namespace is supported on a given platform.
+
+The granularity is kept at a large scale to have large blocks of functionality controlled
+by a single `IsSupported`.
+
+#### Intrinsic granularity
+
+In general individual intrinsic will be chosen to be fine grained.  These will generally correspond to a single assembly instruction.
+
+#### Intrinsics & Crossgen
+
+For any intrinsic which may not be supported on all variants of a platform.  Crossgen
+Method compilation must be trapped, so that the JIT is forced to generate optimal platform dependent 
+code at runtime.
+
+#### API review
+
+Intrinsics will extend the API of CoreCLR.  They will need to follow standard API review practices.
+
+### Choice of Arm64 naming conventions
+
+#### Namespaces
+
+`ARM32` and `ARM64` will follow similar namespace conventions.
+
++ `System.Runtime.Intrinsics` is used for type definitions useful across multiple platforms
++ `System.Runtime.Intrinsics.Arm` is used for enumerations shared across `ARM32` and `ARM64` platforms
++ `System.Runtime.Intrinsics.Arm.Base` is recommended for functionality included in all `ARM32` and `ARM64` platforms
+  + Do we need this or could we assume `ARMv8` is minimum supported platform?
+  + Should we assume `ARMv7` or `ARMv5` is minimum supported platform?
++ `System.Runtime.Intrinsics.Arm.ARMv8` is used for `ARMv8` added functionality included in both `ARM32` and `ARM64` platforms 
++ `System.Runtime.Intrinsics.Arm.ARMv8.Arm64` is used for `ARMv8` functionality only included in `ARM64` platform
++ `System.Runtime.Intrinsics.Arm.ARMv8_1....` is used for `ARMv8.1` functionality.  
+  + Use similar sub namespace names for only included in `ARM64` platform ...
++ `System.Runtime.Intrinsics.Arm.XXX` is used for optional features `CRC32`, `AES`, `SHA1` ... 
+  + All extensions which were not required when they were initially introduced
+  + For example, `CRC32` became required with `ARMv8.1`, but since it was introduced as optional in `ARMv8` it would go in its own namespace
+  
+If an extension is only supported on `Arm32`, there is no reason to add a sub namespace.
+
++ `System.Runtime.Intrinsics.Arm.ARMv7` is used for `ARMv7` added functionality not included in `Base`
+
+If an extension is only supported on `Arm64`, there is no reason to add a sub namespace.  Hypothetically:
+
++ `System.Runtime.Intrinsics.Arm.ARMv10` is used for `ARMv10` added functionality (Hypothetically `Arm64` only)
```

Comment:
My impression is that ISA extensions evolve independently from ISA specification. There is no reason to have ARMv10 intrinsics as long as there is no new ISA extensions. Arm follows Intel in giving names or acronyms to it's ISA extensions.

Reply:
ARMv10 is a hypothetical architectural extension, where ARM32 supported completed dropped.  It is purely a grammatical construction for example.

> My impression is that ISA extensions evolve independently from ISA specification. There is no reason to have ARMv10 intrinsics as long as there is no new ISA extensions. Arm follows Intel in giving names or acronyms to it's ISA extensions.

This statement does not make much sense to me.

ARM is independent from Intel.. It does not follow Intel any more than Intel follows Arm.  

Regarding naming conventions.  Sometimes Architectural Extensions use names like SVE, other times they use official names like ARMv8.1.